src/main/kotlin/com/openlattice/organizations/ExternalDatabaseManagementService.kt

```diff
@@ -0,0 +1,369 @@
+    private val organizationExternalDatabaseTables: IMap<UUID, OrganizationExternalDatabaseTable> = hazelcastInstance.getMap(HazelcastMap.ORGANIZATION_EXTERNAL_DATABASE_TABLE.name)
+    private val securableObjectTypes: IMap<AclKey, SecurableObjectType> = hazelcastInstance.getMap(HazelcastMap.SECURABLE_OBJECT_TYPES.name)
+    private val aces: IMap<AceKey, AceValue> = hazelcastInstance.getMap(HazelcastMap.PERMISSIONS.name)
+    private val logger = LoggerFactory.getLogger(ExternalDatabaseManagementService::class.java)
+
+    /*CREATE*/
+    fun createOrganizationExternalDatabaseTable(orgId: UUID, table: OrganizationExternalDatabaseTable): UUID {
+        val principal = Principals.getCurrentUser()
+        Principals.ensureUser(principal)
+
+        val tableFQN = FullQualifiedName(orgId.toString(), table.name)
+        checkState(organizationExternalDatabaseTables.putIfAbsent(table.id, table) == null,
+                "OrganizationExternalDatabaseTable ${tableFQN.fullQualifiedNameAsString} already exists")
+        aclKeyReservations.reserveIdAndValidateType(table, tableFQN::getFullQualifiedNameAsString)
+
+        val tableAclKey = AclKey(orgId, table.id)
+        authorizationManager.setSecurableObjectType(tableAclKey, SecurableObjectType.OrganizationExternalDatabaseTable)
+        authorizationManager.addPermission(tableAclKey, principal, EnumSet.allOf(Permission::class.java))
+        //eventBus?
+
+        return table.id
+    }
+
+    fun createOrganizationExternalDatabaseColumn(orgId: UUID, column: OrganizationExternalDatabaseColumn): UUID {
+        val principal = Principals.getCurrentUser()
+        Principals.ensureUser(principal)
+
+        checkState(organizationExternalDatabaseTables[column.tableId] != null,
+                "OrganizationExternalDatabaseColumn ${column.name} belongs to a table that does not exist")
+        val columnFQN = FullQualifiedName(column.tableId.toString(), column.name)
+        checkState(organizationExternalDatabaseColumns.putIfAbsent(column.id, column) == null,
+                "OrganizationExternalDatabaseColumn ${columnFQN.fullQualifiedNameAsString} already exists")
+        aclKeyReservations.reserveIdAndValidateType(column, columnFQN::getFullQualifiedNameAsString)
+
+        val columnAclKey = AclKey(orgId, column.tableId, column.id)
+        authorizationManager.setSecurableObjectType(columnAclKey, SecurableObjectType.OrganizationExternalDatabaseColumn)
+        authorizationManager.addPermission(columnAclKey, principal, EnumSet.allOf(Permission::class.java))
+
+        return column.id
+    }
+
+    /*GET*/
+    fun getOrganizationIds(): Set<UUID> {
+        val orgIds = mutableSetOf<UUID>()
+        hds.connection.createStatement().use {
+            val rs = it.executeQuery("SELECT id FROM organizations")
+            while (rs.next()) {
+                orgIds.add(organizationId(rs))
+            }
+        }
```

Comment:
BTW: we already have MapStores, that store organization ids as their keys: `OrganizationAssemblyMapstore`, `OrganizationAppsMapstore`, `OrganizationDescriptionsMapstore`, `OrganizationEmailDomainsMapstore`..

Reply:
Oh ^^^ this about mapstores

But yeah, I think I prefer using the BasePGI, just so that we have consistency around any supports/metrics/internal logic we build for ourselves around querying postgres. I think the custom logic there is relatively minimal right now, but it does handle connection timeouts and retries and closing result sets on exceptions and all that stuff, and I can imagine that support getting built out further in the future.

Reply:
Cool, I have a whole bunch of this logic sprinkled about, I'll convert it all to BPGI